toolchain/check/handle_struct.cpp

```diff
@@ -52,6 +52,64 @@ auto HandleStructFieldValue(Context& context, Parse::Node parse_node) -> bool {
   return true;
 }
 
+// Goes from the StructFieldValue node pointing to the `=` in a struct literal
+// like `{.a = 1}`, to the name node `a` in its children.
+static auto StructFieldValueToName(Context& context,
+                                   Parse::Node field_value_node)
+    -> Parse::Node {
+  // TODO: Will be easier after #3393.
+  CARBON_CHECK(context.parse_tree().node_kind(field_value_node) ==
+               Parse::NodeKind::StructFieldValue);
+  auto children = context.parse_tree().children(field_value_node);
+  auto struct_field_designator_iter = children.begin();
+  ++struct_field_designator_iter;
+  CARBON_CHECK(context.parse_tree().node_kind(*struct_field_designator_iter) ==
+               Parse::NodeKind::StructFieldDesignator);
+  children = context.parse_tree().children(*struct_field_designator_iter);
+  return *children.begin();
+}
+
+static auto HasDuplicateNamesError(Context& context,
+                                   SemIR::InstBlockId type_block_id) -> bool {
+  auto& sem_ir = context.sem_ir();
+  auto fields = sem_ir.inst_blocks().Get(type_block_id);
+  llvm::DenseMap<SemIR::NameId, Parse::Node> names;
+  auto& insts = sem_ir.insts();
+  for (SemIR::InstId field_inst_id : fields) {
+    auto field_inst = insts.GetAs<SemIR::StructTypeField>(field_inst_id);
+    auto [it, added] =
+        names.insert({field_inst.name_id, field_inst.parse_node});
+    if (!added) {
+      // FIXME: Should this be renamed `MemberNameDuplicate`, and possibly
+      // made more widely available?
+      CARBON_DIAGNOSTIC(StructNameDuplicate, Error,
+                        "Member of {0} with a duplicated name `{1}`.",
```

Comment:
I think the style we've been using for diagnostics recently would drop the "a" here.

```suggestion
                        "Member of {0} with duplicated name `{1}`.",
```
I think we could even shorten this to
```suggestion
                        "Duplicated field name `{1}` in {0}.",
```
("Member" -> "field" also seems like it'd make the diagnostic a bit more precise.)